### PR TITLE
feat: Implement edit dataset frontend component (#7061)

### PR DESCRIPTION
**Pull request by frano-m**
## Reason for Change

- #7061.

## Changes

- Added edit dataset functionality from the private or revision collection page as per [mocks](https://www.figma.com/design/jEv0ZWQd0Q53aFl9nNRbBy/Portal?node-id=1250-8164&t=lBHkJBLVfa6QxB5Y-4) including dialog, and error.
- Added form functionality to track changes in title, and execute PATCH of updates on click on save.

~## Note~

~`PATCH` requests on RDEV are currently not working (see https://czi-tech.atlassian.net/browse/ONCALL-824?atlOrigin=eyJpIjoiNWVkODFjNzg1Njk0NDkzYWIwNWM5YzhiMTQ4MzkzZTkiLCJwIjoiamlyYS1zbGFjay1pbnQifQ).~

### Steps for testing private collection.

1. Create `private` collection and add new dataset.

**Expect** - the "more" menu, with option "Edit Dataset" should not be enabled unless the dataset is finished loading, and has not failed.

2. The dataset is loaded and valid.

**Expect** - the "more" menu, with option "Edit Dataset" should be enabled.

3. Click on the "Edit Dataset" menu item.

**Expect** - the "Edit Dataset" dialog should open. The "Save" button should be disabled (form is not dirty).

4. Delete the text field "Name" value, and press "Save".

**Expect** - the form text field to error with "Name is required".

5. Update the title with "New title 01".

**Expect** - the error to clear.

6. Click "Save".

**Expect** - the dialog to close, the "more" menu to close, and invalidation of the collection i.e. the dataset should update with the new name.

### Steps for testing revision collection.

1. Publish `private` collection.
2. Click on "Start Revision" button.
3. Follow steps 1 - 6 above for `private` collection (rename the dataset title "New title 02").
4. Click on "Publish".
5. Go to Datasets index.

**Expect** - the dataset title to be "New title 02".

## Checklist 🛎️

- [ ] Add product, design, and eng as reviewers for rdev review
- [ ] For UI changes, add screenshots/videos, so the reviewers know what you expect them to see
- [ ] For UI changes, add e2e tests to prevent regressions
- [ ] For UI changes, verify impacted analytics events still work

### CODE REVIEW

File: frontend/src/components/Collections/components/Dataset/components/DownloadDataset/index.tsx [hunk 1]
```diff
@@ -17,7 +18,7 @@ const DownloadDataset: FC<Props> = ({
   isDisabled = false,
   name,
 }) => {
-  const [isOpen, setIsOpen] = React.useState(false);
+  const { onClose, onOpen, open } = useDialog();
```
Comment (frano-m) [inline frontend/src/components/Collections/components/Dataset/components/DownloadDataset/index.tsx:21]: Collection view dialog components now use a basic `useDialog` hook that controls dialog state and functionality i.e. `open`, `onClose` (refer [shared hook](https://github.com/chanzuckerberg/single-cell-data-portal/pull/7234/files#diff-b25d215d4b7fd16cceae5090e40f467bb6b46e838e05cfa1d90c06bb2db15b67R13)).

File: frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/DeleteDataset/index.tsx [hunk 2]
```diff
@@ -12,16 +11,12 @@ const AsyncAlert = loadable(
 );
 
 interface Props {
-  Button?: React.ElementType;
+  Button: ElementType;
```
Comment (frano-m) [inline frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/DeleteDataset/index.tsx:14]: The `DeleteDataset` component is now only used by the dataset row "more" menu. The `Button` prop is now updated to a SDS component, and therefore the `Button` prop is changed from `optional` to `required`, thereby allowing us to remove the BlueprintJs `DefaultButton`.

File: frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/EditDataset/components/EditDatasetFields/index.tsx
Comment (frano-m) [inline frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/EditDataset/components/EditDatasetFields/index.tsx:1]: The `EditDatasetFields` component currently only edits dataset "title"; however can be extended in the future to edit other fields, should it be required.

File: frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/EditDataset/components/EditDatasetFields/utils.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import { Dataset } from "src/common/entities";
+
+/**
+ * Maps dataset to form default values.
+ * @param dataset - Dataset.
+ * @returns dataset mapped to form default values.
+ */
+export function mapDatasetToFormDefaultValues(dataset: Dataset): FieldValues {
+  return {
+    title: dataset.name,
```
Comment (frano-m) [inline frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/EditDataset/components/EditDatasetFields/utils.ts:11]: At present, title is the only field we will be editing!

File: frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/EditDataset/index.tsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import EditDatasetFields from "src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/EditDataset/components/EditDatasetFields";
+
+export default function EditDataset({
+  Button,
+  collectionId,
+  dataset,
+  editDataset,
+  menuProps,
+}: Props): JSX.Element {
+  const { onClose, onOpen, open } = useDialog();
```
Comment (frano-m) [inline frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/EditDataset/index.tsx:23]: New basic `useDialog` hook for Collection view related dialog components.

File: frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/EditDataset/index.tsx [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    menuProps.onClose(); // Close the "more" menu after the dialog closes.
+  }, [menuProps, resetForm]);
+
+  return (
+    <Fragment>
+      <Button onClick={onOpen} />
+      <StyledDialog
+        {...DIALOG_PROPS}
+        onClose={onClose}
+        onSubmit={handleSubmit(
```
Comment (frano-m) [inline frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/EditDataset/index.tsx:43]: Form submission! We provide the action required with the submit, and path parameters associated with the request URL, and callback functions associated with the submit handler.

File: frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/MoreDropdown/components/Menu/index.tsx [hunk 6]
```diff
@@ -1,86 +1,76 @@
...
+  MENU_PROPS,
+} from "src/views/Collection/components/CollectionActions/components/MoreDropdown/components/Menu/constants";
+import { MENU_ITEM_COLOR } from "src/views/Collection/components/CollectionActions/components/MoreDropdown/components/Menu/types";
+import { Icon } from "@czi-sds/components";
+import EditDataset from "src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/EditDataset";
+import { isDeleteDatasetAvailable } from "src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/MoreDropdown/components/Menu/utils";
+import { DROPDOWN_EDIT_DATASET } from "src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/MoreDropdown/components/Menu/constants";
 
-const Menu = ({
+export default function Menu({
```
Comment (frano-m) [inline frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/MoreDropdown/components/Menu/index.tsx:19]: Complete overhaul of the dataset row "more" menu!

Updates include:
- changed FC to preferred "regular" style,
- grouped props by use, and
- use of supporting files such as `types.ts` as per WMG code style.

File: frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/MoreDropdown/components/Menu/index.tsx [hunk 7]
```diff
@@ -1,86 +1,76 @@
...
   collectionId,
-  datasetId = "",
-  isPublished,
-  revisionsEnabled,
-  onUploadFile,
-  isLoading,
-}: Props): JSX.Element => {
-  // A dataset may be deleted if the collection is private, or the dataset has not been previously
-  // published; where the published_at property is used to determine whether the dataset has been previously published.
-  const shouldShowDelete = !revisionsEnabled || !isPublished;
```
Comment (frano-m) [inline frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/MoreDropdown/components/Menu/index.tsx:67]: `shouldShowDelete` logic is now encapsulated in a function [isDeleteDatasetAvailable](https://github.com/chanzuckerberg/single-cell-data-portal/pull/7234/files#diff-2b59f9159123e7170cd4b47f1e0bc5fb828848539c96d5c4d9292423fa59eb7dR27).

File: frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/MoreDropdown/components/Menu/index.tsx [hunk 8]
```diff
@@ -1,86 +1,76 @@
...
-} from "@blueprintjs/core";
-import { IconNames } from "@blueprintjs/icons";
-import styled from "@emotion/styled";
-import DropboxChooser, {
-  Props as ChooserProps,
-} from "src/components/DropboxChooser";
-import DeleteDataset from "../../../DeleteDataset";
-import { Collection } from "src/common/entities";
-
-const DeleteButton = (props: MenuItemProps) => {
```
Comment (frano-m) [inline frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/MoreDropdown/components/Menu/index.tsx:15]: Moved `DeleteButton` and `UpdateButton` etc into main component for readability.

File: frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/MoreDropdown/index.tsx [hunk 9]
```diff
@@ -1,57 +1,35 @@
...
-  datasetId?: string;
-  isPublished: boolean;
-  revisionsEnabled: boolean;
-  onUploadFile: ChooserProps["onUploadFile"];
-  isLoading: boolean;
-  disabled: boolean;
-}
-
-const MoreDropdown = ({
+export default function MoreDropdown({
```
Comment (frano-m) [inline frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/components/MoreDropdown/index.tsx:11]: Complete overhaul of dataset row `MoreDropdown` component!

- changed FC to preferred "regular" style,
- grouped props by use,
- use of supporting files such as types.ts as per WMG code style, and
- replacement of BlueprintJS components with SDS components.

File: frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/index.tsx [hunk 10]
```diff
@@ -243,12 +250,15 @@ const DatasetRow: FC<Props> = ({
...
               <MoreDropdown
                 collectionId={collectionId}
-                datasetId={dataset.id}
-                isPublished={!!dataset.published_at} // Dataset has been published.
-                revisionsEnabled={revisionsEnabled}
-                onUploadFile={onUploadFile}
-                isLoading={isLoading}
+                dataset={dataset}
                 disabled={dataset.tombstone ?? false}
+                menuItemProps={{
```
Comment (frano-m) [inline frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DatsetRow/index.tsx:255]: Menu Item related props are grouped for readability.

File: frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DownloadDataset/index.tsx
Comment (frano-m) [inline frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DownloadDataset/index.tsx:1]: Collection view dialog components now use a basic useDialog hook that controls dialog state and functionality i.e. open, onClose (refer [shared hook](https://github.com/chanzuckerberg/single-cell-data-portal/pull/7234/files#diff-b25d215d4b7fd16cceae5090e40f467bb6b46e838e05cfa1d90c06bb2db15b67R13)).

File: frontend/src/components/common/Button/index.tsx [hunk 11]
```diff
@@ -15,6 +15,7 @@ import {
 import { css, SerializedStyles } from "@emotion/react";
 
 type ButtonProps = RawButtonProps & CommonThemeProps;
+type IconButtonProps = ButtonProps & { open?: boolean };
```
Comment (frano-m) [inline frontend/src/components/common/Button/index.tsx:18]: Added `sdsStyle` `icon` to common Button component - to add `open` state (and therefore "open" color) to button as per [mocks](https://www.figma.com/design/jEv0ZWQd0Q53aFl9nNRbBy/Portal?node-id=1250-8974&t=lBHkJBLVfa6QxB5Y-4). 

This style is used by both Collection view "more" icon button, and Dataset Row "more" icon button (refer [here](https://github.com/chanzuckerberg/single-cell-data-portal/pull/7234/files#diff-d96d3222524eb4739bb176f9768400695b12eb37e0315bdfd1c2236911d46ff1R7) and [here](https://github.com/chanzuckerberg/single-cell-data-portal/pull/7234/files#diff-9e5d5cc8828db3bdc41c98a2435da2bcc3e74cd021c12ca869ccafc3a9460a8b)).

File: frontend/src/components/common/MoreDropdown/index.tsx [hunk 12]
```diff
@@ -1,27 +0,0 @@
-import { Popover, PopoverProps } from "@blueprintjs/core";
```
Comment (frano-m) [inline frontend/src/components/common/MoreDropdown/index.tsx:1]: Removal of no-longer-in-use BlueprintJS component!

File: frontend/src/components/common/MoreDropdown/index.tsx
Comment (frano-m) [inline frontend/src/components/common/MoreDropdown/index.tsx:1]: Removal of no-longer-in-use BlueprintJS component!

File: frontend/src/views/Collection/components/CollectionActions/components/MoreDropdown/components/Menu/components/IconSort/index.tsx [hunk 13]
```diff
@@ -1,20 +1,25 @@
...
-      <path
-        d="M3.70711 0.292893C3.31658 -0.0976311 2.68342 -0.0976311 2.29289 0.292893L0.292893 2.29289C-0.0976311 2.68342 -0.0976311 3.31658 0.292893 3.70711C0.683417 4.09763 1.31658 4.09763 1.70711 3.70711L2 3.41421V10C2 10.5523 2.44772 11 3 11C3.55228 11 4 10.5523 4 10V3.41421L4.29289 3.70711C4.68342 4.09763 5.31658 4.09763 5.70711 3.70711C6.09763 3.31658 6.09763 2.68342 5.70711 2.29289L3.70711 0.292893Z"
-        fill="currentColor"
-      />
-      <path
-        d="M11.7071 9.70711L9.70711 11.7071C9.31658 12.0976 8.68342 12.0976 8.29289 11.7071L6.29289 9.70711C5.90237 9.31658 5.90237 8.68342 6.29289 8.29289C6.68342 7.90237 7.31658 7.90237 7.70711 8.29289L8 8.58579V2C8 1.44771 8.44772 1 9 1C9.55228 1 10 1.44771 10 2V8.58579L10.2929 8.29289C10.6834 7.90237 11.3166 7.90237 11.7071 8.29289C12.0976 8.68342 12.0976 9.31658 11.7071 9.70711Z"
-        fill="currentColor"
-      />
-    </svg>
+    <SvgIcon className={className} color="inherit">
```
Comment (frano-m) [inline frontend/src/views/Collection/components/CollectionActions/components/MoreDropdown/components/Menu/components/IconSort/index.tsx:6]: Custom icon `IconSort` updated to allow for style overrides.

File: frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DownloadDataset/index.tsx
Comment (frano-m) [inline frontend/src/components/Collection/components/CollectionDatasetsGrid/components/Row/DownloadDataset/index.tsx:1]: Collection view dialog components now use a basic useDialog hook that controls dialog state and functionality i.e. open, onClose (refer [shared hook](https://github.com/chanzuckerberg/single-cell-data-portal/pull/7234/files#diff-b25d215d4b7fd16cceae5090e40f467bb6b46e838e05cfa1d90c06bb2db15b67R13)).

File: frontend/src/views/Collection/common/style.ts
Comment (frano-m) [inline frontend/src/views/Collection/common/style.ts:1]: Commonly used Collection view dialog component styles.

File: frontend/src/views/Collection/components/CollectionActions/components/MoreDropdown/components/Menu/index.tsx
Comment (frano-m) [inline frontend/src/views/Collection/components/CollectionActions/components/MoreDropdown/components/Menu/index.tsx:1]: Collection view and Dataset Row "more" dropdown components share common styled menu, styled menu items, and icons (see [styled components](https://github.com/chanzuckerberg/single-cell-data-portal/pull/7234/files#diff-94a9c87997fcabfd4262bddde23867db83193799a1cf741b0b25d88bf1a79e08).

File: frontend/src/views/Collection/hooks/useCollectionDatasetForm/useCollectionDatasetForm.ts [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+  DEFAULT_DIRTY_FIELDS,
+  DEFAULT_FIELD_ERRORS,
+  DEFAULT_FIELD_VALUES,
+} from "src/views/Collection/hooks/useCollectionDatasetForm/constants";
+
+/**
+ * Collection dataset form functionality for collection dataset.
+ * @returns collection dataset form functionality.
+ */
+export function useCollectionDatasetForm(): UseCollectionDatasetForm {
```
Comment (frano-m) [inline frontend/src/views/Collection/hooks/useCollectionDatasetForm/useCollectionDatasetForm.ts:35]: Collection view hook handling form related actions i.e. submit with form validation, error handling, dirty fields ...

File: frontend/src/views/Collection/hooks/useEditCollectionDataset/useEditCollectionDataset.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+  PathParams,
+  SubmitOptions,
+} from "src/views/Collection/hooks/useCollectionDatasetForm/types";
+import { useCallback } from "react";
+import { useCollectionDatasetForm } from "src/views/Collection/hooks/useCollectionDatasetForm/useCollectionDatasetForm";
+
+/**
+ * Edit functionality for collection dataset.
+ */
+export function useEditCollectionDataset(): UseEditCollectionDataset {
```
Comment (frano-m) [inline frontend/src/views/Collection/hooks/useEditCollectionDataset/useEditCollectionDataset.ts:14]: Collection view hook for the editing of a dataset (title).

File: frontend/src/common/queries/collections.ts [hunk 16]
```diff
@@ -416,6 +416,63 @@ export function usePublishCollection() {
...
+ */
+export function useEditDataset() {
+  const queryClient = useQueryClient();
+  return useMutation(editDataset, {
+    onSuccess: async (_, variables): Promise<void> => {
+      await queryClient.invalidateQueries([
+        USE_COLLECTION,
+        variables.collectionId,
+      ]);
+      await queryClient.invalidateQueries(
```
Comment (frano-m) [inline frontend/src/common/queries/collections.ts:468]: Maintaining Datasets Index with a background refetch.

**Review by kaloster — APPROVED** [review]
Thank you for a stellar PR! LGTM 🚀

**Pull request merged**